#01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

#02. Importing Data

In [2]:
# Import the pandas, NumPy, and os libraries
df = pd.read_csv(r'/Users/thomasradio/Documents/Jupyter/230620 Instacart Basket Analysis/02 Data/Original Data/orders.csv', index_col = False)

In [3]:
# Turn project folder path into a string
path = r'/Users/thomasradio/Documents/Jupyter/230621 Tom Radio Achievement 4 Project'

In [4]:
# Check path
path

'/Users/thomasradio/Documents/Jupyter/230621 Tom Radio Achievement 4 Project'

In [5]:
# Import the “products.csv” data set into notebook using the os library
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [7]:
# Import the “orders_wrangled.csv” data set into notebook using the os library
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   user_order_number       int64  
 5   orders_day_of_week      int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


In [9]:
# Drop "Unnamed: 0"
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [10]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   user_order_number       int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [11]:
df_ords

order_id  user_id eval_set  user_order_number  orders_day_of_week  \
0         2539329        1    prior                  1                   2   
1         2398795        1    prior                  2                   3   
2          473747        1    prior                  3                   3   
3         2254736        1    prior                  4                   4   
4          431534        1    prior                  5                   4   
...           ...      ...      ...                ...                 ...   
3421078   2266710   206209    prior                 10                   5   
3421079   1854736   206209    prior                 11                   4   
3421080    626363   206209    prior                 12                   1   
3421081   2977660   206209    prior                 13                   1   
3421082    272231   206209    train                 14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [12]:
# Drop eval_set
df_ords = df_ords.drop(columns = ['eval_set'])

In [13]:
df_ords

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
1         2398795        1                  2                   3   
2          473747        1                  3                   3   
3         2254736        1                  4                   4   
4          431534        1                  5                   4   
...           ...      ...                ...                 ...   
3421078   2266710   206209                 10                   5   
3421079   1854736   206209                 11                   4   
3421080    626363   206209                 12                   1   
3421081   2977660   206209                 13                   1   
3421082    272231   206209                 14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

# Save and checkpoint

In [14]:
# Create a dataframe
df_test = pd.DataFrame()

In [15]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
# Convert mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# Create subset containg null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
# Find the number of rows and columns
df_prods.shape

(49693, 5)

In [23]:
# Create subset containg non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
df_prods_clean.shape

(49677, 5)

# Save and checkpoint

In [25]:
# Look for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
# Create new dataframe that doesn't contain duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

# Save and checkpoint

In [30]:
# Export as "products_checked"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Save and checkpoint

# Task 4.5

# 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

    Tip: Keep an eye on min and max values!


In [31]:
df_ords.describe()

order_id       user_id  user_order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06       3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05       1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04       1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00       1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04       5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05       1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05       2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05       1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## order_id min 1 / max 3.42
## user_id max 2
## user_order_number max 1
## order_hour_of_day max 2.3
## days_since_prior_order max 3

# 3. Check for mixed-type data in your df_ords dataframe.

In [32]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## No columns containing mixed data types found.

# 5. Run a check for missing values in your df_ords dataframe.

    In a markdown cell, report your findings and propose an explanation for any missing values you find.


In [33]:
# Find missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
user_order_number              0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## days_since_prior_order contains 206,209 missing values
## These values are missing because it is the customer's first ever order.

# 6. Address the missing values using an appropriate method.

    In a markdown cell, explain why you used your method of choice.


## I chose to leave the "NaN" entries as is. There is no need to flag them, imputing a value would make them incorrect, and you don't want to remove them because they aren't wrong.

# 7. Run a check for duplicate values in your df_ords data.

    In a markdown cell, report your findings and propose an explanation for any duplicate values you find.


In [34]:
# Look for full duplicates
df_dups = df_ords[df_ords.duplicated()]

In [35]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, user_order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## No duplicates found.

# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [36]:
# Export as "orders_checked"
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

# Save and checkpoint

In [37]:
df_ords

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
1         2398795        1                  2                   3   
2          473747        1                  3                   3   
3         2254736        1                  4                   4   
4          431534        1                  5                   4   
...           ...      ...                ...                 ...   
3421078   2266710   206209                 10                   5   
3421079   1854736   206209                 11                   4   
3421080    626363   206209                 12                   1   
3421081   2977660   206209                 13                   1   
3421082    272231   206209                 14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

# DELETE BACK TO HERE

In [38]:
df_ords

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
1         2398795        1                  2                   3   
2          473747        1                  3                   3   
3         2254736        1                  4                   4   
4          431534        1                  5                   4   
...           ...      ...                ...                 ...   
3421078   2266710   206209                 10                   5   
3421079   1854736   206209                 11                   4   
3421080    626363   206209                 12                   1   
3421081   2977660   206209                 13                   1   
3421082    272231   206209                 14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]